File: python/src/lightcone_sdk/domain/market/convert.py
"""Market wire-to-domain conversion."""

from typing import Optional
from . import (
    Market, MarketValidationError, Status, Outcome, ConditionalToken, DepositAsset,
    DepositAssetPair, GlobalDepositAsset, TokenMetadata, sort_by_display_priority,
)
from ...error import SdkError
from ..orderbook import OrderBookPair
from .wire import GlobalDepositAssetWire, MarketWire


def _resolve_icon_urls(
    low: str | None,
    medium: str | None,
    high: str | None,
) -> tuple[str, str, str] | None:
    """Resolve three icon URL quality variants with cross-fallback.
    Returns None if all three are missing/empty."""
    any_url = low or medium or high
    if not any_url:
        return None
    return (
        low or medium or high or any_url,
        medium or low or high or any_url,
        high or medium or low or any_url,
    )


def _parse_status(s: Optional[str]) -> Status:
    if s is None:
        return Status.PENDING
    return Status.from_str(s)


def validation_errors_from_wire(wire: MarketWire) -> list[str]:
    errors: list[str] = []

    if not wire.slug:
        errors.append("Missing slug")
    if not wire.market_name:
        errors.append("Missing name")
    if not wire.description:
        errors.append("Missing description")
    if not wire.definition:
        errors.append("Missing definition")
    if _resolve_icon_urls(wire.icon_url_low, wire.icon_url_medium, wire.icon_url_high) is None:
        errors.append("Missing thumbnail image")
    if _resolve_icon_urls(wire.banner_image_url_low, wire.banner_image_url_medium, wire.banner_image_url_high) is None:
        errors.append("Missing banner image")
    if wire.market_status and wire.market_status not in {"Pending", "Active", "Resolved", "Cancelled"}:
        errors.append("Invalid status")

    if not errors:
        return []

    identifier = wire.market_pubkey or str(wire.market_id)
    return [f"Market validation errors ({identifier}): {', '.join(errors)}"]


def market_from_wire(wire: MarketWire) -> Market:
    """Convert a MarketWire to a Market domain type."""
    outcomes = []
    for o in wire.outcomes:
        outcome_icons = _resolve_icon_urls(o.icon_url_low, o.icon_url_medium, o.icon_url_high)
        outcomes.append(Outcome(
            index=o.index,
            name=o.name,
            icon_url_low=outcome_icons[0] if outcome_icons else "",
            icon_url_medium=outcome_icons[1] if outcome_icons else "",
            icon_url_high=outcome_icons[2] if outcome_icons else "",
        ))

    conditional_tokens: list[ConditionalToken] = []
    token_metadata: dict[str, TokenMetadata] = {}
    deposit_assets: list[DepositAsset] = []
    orderbook_ids: list[str] = []

    for da in wire.deposit_assets:
        da_symbol = da.token_symbol or da.symbol
        da_icons = _resolve_icon_urls(da.icon_url_low, da.icon_url_medium, da.icon_url_high)
        deposit_assets.append(DepositAsset(
            id=da.id,
            market_pda=da.market_pubkey,
            deposit_asset=da.deposit_asset,
            num_outcomes=da.num_outcomes,
            name=da.display_name,
            symbol=da_symbol,
            description=da.description,
            decimals=da.decimals,
            icon_url_low=da_icons[0] if da_icons else "",
            icon_url_medium=da_icons[1] if da_icons else "",
            icon_url_high=da_icons[2] if da_icons else "",
        ))

        for ct in da.conditional_mints:
            ct_name = ct.outcome
            ct_symbol = ct.short_symbol or ct.symbol
            ct_icons = _resolve_icon_urls(ct.icon_url_low, ct.icon_url_medium, ct.icon_url_high)
            if ct_icons is None and da_icons is not None:
                ct_icons = da_icons
            conditional_tokens.append(ConditionalToken(
                pubkey=ct.token_address,
                outcome_index=ct.outcome_index,
                id=ct.id,
                outcome=ct.outcome,
                deposit_asset=da.deposit_asset,
                deposit_symbol=da_symbol,
                name=ct_name,
                symbol=ct_symbol,
                description=ct.description,
                decimals=ct.decimals,
                icon_url_low=ct_icons[0] if ct_icons else "",
                icon_url_medium=ct_icons[1] if ct_icons else "",
                icon_url_high=ct_icons[2] if ct_icons else "",
            ))
            if ct.token_address:
                token_metadata[ct.token_address] = TokenMetadata(
                    pubkey=ct.token_address,
                    symbol=ct_symbol,
                    decimals=ct.decimals,
                    icon_url_low=ct_icons[0] if ct_icons else "",
                    icon_url_medium=ct_icons[1] if ct_icons else "",
                    icon_url_high=ct_icons[2] if ct_icons else "",
                    name=ct_name,
                )

    # Build a lookup from pubkey to ConditionalToken for orderbook pairs.
    ct_by_pubkey: dict[str, ConditionalToken] = {ct.pubkey: ct for ct in conditional_tokens}

    orderbook_pairs = []
    for ob in wire.orderbooks:
        base_ct = ct_by_pubkey.get(ob.base_token, ConditionalToken(pubkey=ob.base_token, outcome_index=0))
        quote_ct = ct_by_pubkey.get(ob.quote_token, ConditionalToken(pubkey=ob.quote_token, outcome_index=0))
        orderbook_pairs.append(OrderBookPair(
            id=ob.id,
            market_pubkey=ob.market_pubkey or wire.market_pubkey,
            orderbook_id=ob.orderbook_id,
            base=base_ct,
            quote=quote_ct,
            outcome_index=ob.outcome_index,
            tick_size=int(ob.tick_size) if ob.tick_size is not None else 0,
            total_bids=ob.total_bids,
            total_asks=ob.total_asks,
            last_trade_price=ob.last_trade_price,
            last_trade_time=ob.last_trade_time,
            active=ob.active,
        ))
        if ob.orderbook_id:
            orderbook_ids.append(ob.orderbook_id)

    deposit_asset_pairs = sort_by_display_priority(
        _derive_deposit_asset_pairs(deposit_assets, orderbook_pairs)
    )

    if not deposit_asset_pairs:
        identifier = wire.market_pubkey or str(wire.market_id)
        raise MarketValidationError(
            f"Market validation errors ({identifier}): Missing deposit asset pairs",
            ["Missing deposit asset pairs"],
        )

    banner_icons = _resolve_icon_urls(
        wire.banner_image_url_low, wire.banner_image_url_medium, wire.banner_image_url_high,
    )
    market_icons = _resolve_icon_urls(
        wire.icon_url_low, wire.icon_url_medium, wire.icon_url_high,
    )

    return Market(
        id=wire.market_id,
        pubkey=wire.market_pubkey,
        name=wire.market_name,
        banner_image_url_low=banner_icons[0] if banner_icons else "",
        banner_image_url_medium=banner_icons[1] if banner_icons else "",
        banner_image_url_high=banner_icons[2] if banner_icons else "",
        icon_url_low=market_icons[0] if market_icons else "",
        icon_url_medium=market_icons[1] if market_icons else "",
        icon_url_high=market_icons[2] if market_icons else "",
        featured_rank=wire.featured_rank,
        volume=wire.volume or "0",
        slug=wire.slug or "",
        status=_parse_status(wire.market_status),
        created_at=wire.created_at,
        activated_at=wire.activated_at,
        settled_at=wire.settled_at,
        resolution=wire.resolution,
        description=wire.description or "",
        definition=wire.definition or "",
        category=wire.category,
        tags=wire.tags,
        deposit_assets=deposit_assets,
        deposit_asset_pairs=deposit_asset_pairs,
        conditional_tokens=conditional_tokens,
        outcomes=outcomes,
        orderbook_pairs=orderbook_pairs,
        orderbook_ids=orderbook_ids,
        token_metadata=token_metadata,
    )


def _derive_deposit_asset_pairs(
    deposit_assets: list[DepositAsset],
    orderbook_pairs: list[OrderBookPair],
) -> list[DepositAssetPair]:
    """Derive unique base/quote deposit-asset pairs.

    Deduplicated by ``(base_pubkey, quote_pubkey)``; orderbook pairs whose
    base or quote deposit asset is not present in ``deposit_assets`` are
    skipped. Order is not guaranteed.
    """
    seen: dict[tuple[str, str], DepositAssetPair] = {}
    for pair in orderbook_pairs:
        base = next(
            (a for a in deposit_assets if a.deposit_asset == pair.base.deposit_asset),
            None,
        )
        quote = next(
            (a for a in deposit_assets if a.deposit_asset == pair.quote.deposit_asset),
            None,
        )
        if base is None or quote is None:
            continue
        key = (base.deposit_asset, quote.deposit_asset)
        if key not in seen:
            seen[key] = DepositAssetPair(
                id=f"{base.deposit_asset}-{quote.deposit_asset}",
                base=base,
                quote=quote,
            )
    return list(seen.values())


def global_deposit_asset_from_wire(
    wire: GlobalDepositAssetWire,
) -> GlobalDepositAsset:
    """Convert a ``GlobalDepositAssetWire`` to a ``GlobalDepositAsset``.

    Raises ``SdkError`` with a rendered multi-error message when required
    fields (``display_name``, ``symbol``, ``icon_url_low``, ``decimals``) are
    missing on the wire payload.
    """
    errors: list[str] = []

    if wire.display_name is None:
        errors.append(f"Missing display name: {wire.mint}")
    if wire.symbol is None:
        errors.append(f"Missing symbol: {wire.mint}")
    global_icons = _resolve_icon_urls(wire.icon_url_low, wire.icon_url_medium, wire.icon_url_high)
    if global_icons is None:
        errors.append(f"Missing icon URL: {wire.mint}")
    if wire.decimals is None:
        errors.append(f"Missing decimals: {wire.mint}")

    if errors:
        rendered = "\n".join(f"  - {error}" for error in errors)
        raise SdkError(
            f"Token validation errors ({wire.mint}):\n{rendered}"
        )

    # Re-resolve after validation (global_icons may be None if error was appended)
    resolved_icons = _resolve_icon_urls(wire.icon_url_low, wire.icon_url_medium, wire.icon_url_high)

    return GlobalDepositAsset(
        id=wire.id,
        deposit_asset=wire.mint,
        name=wire.display_name or "",
        symbol=wire.symbol or "",
        description=wire.description,
        decimals=wire.decimals or 0,
        icon_url_low=resolved_icons[0] if resolved_icons else "",
        icon_url_medium=resolved_icons[1] if resolved_icons else "",
        icon_url_high=resolved_icons[2] if resolved_icons else "",
        whitelist_index=wire.whitelist_index,
        active=wire.active,
    )

File: rust/src/domain/market/convert.rs
//! Conversion: MarketResponse → Market (TryFrom + validation).

use super::outcome;
use super::resolve_icon_urls;
use super::tokens;
use super::tokens::sort_by_display_priority;
use super::wire;
use super::{Market, Status, ValidationError};
use crate::domain::orderbook;
use crate::shared::PubkeyStr;
use rust_decimal::Decimal;
use std::collections::HashMap;

impl TryFrom<wire::MarketResponse> for Market {
    type Error = ValidationError;

    fn try_from(source: wire::MarketResponse) -> Result<Self, Self::Error> {
        let mut errors: Vec<ValidationError> = Vec::new();
        let market_pubkey = source.market_pubkey.clone();

        // Validate outcomes
        let mut outcomes: Vec<outcome::Outcome> = Vec::new();
        for o in source.outcomes.clone() {
            match o.try_into() {
                Ok(validated) => outcomes.push(validated),
                Err(err) => errors.push(ValidationError::Outcome(err)),
            }
        }

        // Validate tokens
        let mut token_metadata: HashMap<PubkeyStr, tokens::TokenMetadata> = HashMap::new();
        let mut deposit_assets: Vec<tokens::DepositAsset> = Vec::new();
        let mut conditional_tokens: Vec<tokens::ConditionalToken> = Vec::new();

        for da in source.deposit_assets.clone() {
            match tokens::ValidatedTokens::try_from(da) {
                Ok(validated) => {
                    deposit_assets.push(validated.token);
                    conditional_tokens.extend(validated.conditionals);
                    token_metadata.extend(validated.metadata);
                }
                Err(err) => errors.push(ValidationError::Token(err)),
            }
        }

        deposit_assets.sort_by(|a, b| a.symbol.cmp(&b.symbol));
        conditional_tokens.sort_by(|a, b| {
            use tokens::Token;
            a.symbol().cmp(b.symbol())
        });

        // Validate orderbooks
        let mut orderbook_pairs = Vec::new();
        for book in source.orderbooks.clone() {
            match orderbook::OrderBookPair::try_from((book, &conditional_tokens)) {
                Ok(pair) => orderbook_pairs.push(pair),
                Err(err) => errors.push(ValidationError::OrderBook(err)),
            }
        }

        let slug = source.slug.clone().unwrap_or_else(|| {
            errors.push(ValidationError::MissingSlug);
            String::new()
        });
        let name = source.market_name.clone().unwrap_or_else(|| {
            errors.push(ValidationError::MarketNameMissing);
            String::new()
        });
        let status = Status::from_str(&source.market_status).unwrap_or_else(|| {
            errors.push(ValidationError::InvalidStatus);
            Status::Pending
        });
        let description = source.description.clone().unwrap_or_else(|| {
            errors.push(ValidationError::MissingDescription);
            String::new()
        });
        let definition = source.definition.clone().unwrap_or_else(|| {
            errors.push(ValidationError::MissingDefinition);
            String::new()
        });
        let (icon_url_low, icon_url_medium, icon_url_high) = resolve_icon_urls(
            source.icon_url_low.clone(),
            source.icon_url_medium.clone(),
            source.icon_url_high.clone(),
        )
        .unwrap_or_else(|| {
            errors.push(ValidationError::MissingIconUrl);
            (String::new(), String::new(), String::new())
        });
        let (banner_image_url_low, banner_image_url_medium, banner_image_url_high) =
            resolve_icon_urls(
                source.banner_image_url_low.clone(),
                source.banner_image_url_medium.clone(),
                source.banner_image_url_high.clone(),
            )
            .unwrap_or_else(|| {
                errors.push(ValidationError::MissingBannerUrl);
                (String::new(), String::new(), String::new())
            });

        let deposit_asset_pairs = sort_by_display_priority(&derive_deposit_asset_pairs(
            &deposit_assets,
            &orderbook_pairs,
        ));

        if deposit_asset_pairs.is_empty() {
            errors.push(ValidationError::MissingDepositAssetPairs);
        }

        if !errors.is_empty() {
            return Err(ValidationError::Multiple(market_pubkey, errors));
        }

        Ok(Market {
            id: source.market_id,
            pubkey: source.market_pubkey.into(),
            volume: Decimal::ZERO,
            featured_rank: source.featured_rank,
            slug,
            name,
            status,
            created_at: source.created_at,
            activated_at: source.activated_at,
            settled_at: source.settled_at,
            resolution: source.resolution,
            description,
            definition,
            tags: source.tags.unwrap_or_default(),
            outcomes,
            icon_url_low,
            icon_url_medium,
            icon_url_high,
            banner_image_url_low,
            banner_image_url_medium,
            banner_image_url_high,
            category: source.category,
            orderbook_ids: orderbook_pairs
                .iter()
                .map(|p| p.orderbook_id.clone())
                .collect(),
            orderbook_pairs,
            deposit_assets,
            deposit_asset_pairs,
            conditional_tokens,
            token_metadata,
        })
    }
}

/// Derive unique base/quote deposit-asset pairs across the market's orderbook
/// pairs. Deduplicated by `(base_pubkey, quote_pubkey)`; orderbook pairs whose
/// base or quote deposit asset is not present in `deposit_assets` are skipped.
fn derive_deposit_asset_pairs(
    deposit_assets: &[tokens::DepositAsset],
    orderbook_pairs: &[orderbook::OrderBookPair],
) -> Vec<tokens::DepositAssetPair> {
    let mut seen: HashMap<(PubkeyStr, PubkeyStr), tokens::DepositAssetPair> = HashMap::new();

    for pair in orderbook_pairs {
        let base = deposit_assets
            .iter()
            .find(|asset| asset.deposit_asset == pair.base.deposit_asset);
        let quote = deposit_assets
            .iter()
            .find(|asset| asset.deposit_asset == pair.quote.deposit_asset);

        if let (Some(base), Some(quote)) = (base, quote) {
            let key = (base.deposit_asset.clone(), quote.deposit_asset.clone());
            seen.entry(key).or_insert_with(|| tokens::DepositAssetPair {
                id: format!("{}-{}", base.deposit_asset, quote.deposit_asset),
                base: base.clone(),
                quote: quote.clone(),
            });
        }
    }

    seen.into_values().collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;

    fn minimal_market_response() -> wire::MarketResponse {
        wire::MarketResponse {
            market_name: Some("Test Market".to_string()),
            slug: Some("test-market".to_string()),
            description: Some("Description".to_string()),
            definition: Some("Definition".to_string()),
            outcomes: vec![wire::OutcomeResponse {
                index: 0,
                name: "Yes".to_string(),
                icon_url_low: Some("https://example.com/yes_low.png".to_string()),
                icon_url_medium: Some("https://example.com/yes_medium.png".to_string()),
                icon_url_high: Some("https://example.com/yes_high.png".to_string()),
            }],
            banner_image_url_low: Some("https://example.com/banner_low.png".to_string()),
            banner_image_url_medium: Some("https://example.com/banner_medium.png".to_string()),
            banner_image_url_high: Some("https://example.com/banner_high.png".to_string()),
            icon_url_low: Some("https://example.com/icon_low.png".to_string()),
            icon_url_medium: Some("https://example.com/icon_medium.png".to_string()),
            icon_url_high: Some("https://example.com/icon_high.png".to_string()),
            category: None,
            tags: None,
            featured_rank: None,
            market_pubkey: "mkt123".to_string(),
            market_id: 1,
            oracle: "oracle".to_string(),
            question_id: "q1".to_string(),
            condition_id: "c1".to_string(),
            market_status: "Active".to_string(),
            resolution: None,
            created_at: Utc::now(),
            activated_at: None,
            settled_at: None,
            deposit_assets: vec![],
            orderbooks: vec![],
        }
    }

    fn valid_market_response(
        resolution: Option<wire::MarketResolutionResponse>,
    ) -> wire::MarketResponse {
        let mut response = minimal_market_response();
        response.market_status = "Resolved".to_string();
        response.resolution = resolution;
        response.deposit_assets = vec![deposit_asset_response()];
        response.orderbooks = vec![orderbook_response()];
        response
    }

    fn deposit_asset_response() -> wire::DepositAssetResponse {
        wire::DepositAssetResponse {
            display_name: Some("USD Coin".to_string()),
            token_symbol: Some("USDC".to_string()),
            symbol: Some("USDC".to_string()),
            deposit_asset: "USDC".to_string(),
            id: 1,
            market_pubkey: "mkt123".to_string(),
            vault: "vault".to_string(),
            num_outcomes: 2,
            description: None,
            icon_url_low: Some("https://example.com/usdc_low.png".to_string()),
            icon_url_medium: Some("https://example.com/usdc_medium.png".to_string()),
            icon_url_high: Some("https://example.com/usdc_high.png".to_string()),
            metadata_uri: None,
            decimals: Some(6),
            conditional_mints: vec![
                conditional_token_response(10, 0, "yes_mint", "Yes", "YES"),
                conditional_token_response(11, 1, "no_mint", "No", "NO"),
            ],
            created_at: Utc::now(),
        }
    }

    fn conditional_token_response(
        id: i32,
        outcome_index: i16,
        token_address: &str,
        outcome: &str,
        short_symbol: &str,
    ) -> wire::ConditionalTokenResponse {
        wire::ConditionalTokenResponse {
            id,
            outcome_index,
            token_address: token_address.to_string(),
            symbol: Some(short_symbol.to_string()),
            uri: None,
            outcome: Some(outcome.to_string()),
            deposit_symbol: Some("USDC".to_string()),
            short_symbol: Some(short_symbol.to_string()),
            description: None,
            icon_url_low: Some(format!("https://example.com/{short_symbol}_low.png")),
            icon_url_medium: Some(format!("https://example.com/{short_symbol}_medium.png")),
            icon_url_high: Some(format!("https://example.com/{short_symbol}_high.png")),
            metadata_uri: None,
            decimals: Some(6),
            created_at: Utc::now(),
        }
    }

    fn orderbook_response() -> orderbook::wire::OrderbookResponse {
        orderbook::wire::OrderbookResponse {
            id: 1,
            market_pubkey: "mkt123".to_string(),
            orderbook_id: "ob_yes_no".to_string(),
            base_token: "yes_mint".to_string(),
            quote_token: "no_mint".to_string(),
            outcome_index: Some(0),
            tick_size: 1,
            total_bids: 0,
            total_asks: 0,
            last_trade_price: None,
            last_trade_time: None,
            active: true,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }

    fn single_winner_resolution() -> wire::MarketResolutionResponse {
        wire::MarketResolutionResponse {
            kind: wire::MarketResolutionKind::SingleWinner,
            payout_denominator: 1,
            payouts: vec![
                wire::MarketResolutionPayout {
                    outcome_index: 0,
                    payout_numerator: 0,
                },
                wire::MarketResolutionPayout {
                    outcome_index: 1,
                    payout_numerator: 1,
                },
            ],
            single_winning_outcome: Some(1),
        }
    }

    fn scalar_resolution() -> wire::MarketResolutionResponse {
        wire::MarketResolutionResponse {
            kind: wire::MarketResolutionKind::Scalar,
            payout_denominator: 10,
            payouts: vec![
                wire::MarketResolutionPayout {
                    outcome_index: 0,
                    payout_numerator: 7,
                },
                wire::MarketResolutionPayout {
                    outcome_index: 1,
                    payout_numerator: 3,
                },
            ],
            single_winning_outcome: None,
        }
    }

    #[test]
    fn market_response_resolution_helpers_handle_unresolved_scalar_and_winner() {
        let mut response = minimal_market_response();
        assert!(!response.is_resolved());
        assert_eq!(response.single_winning_outcome(), None);
        assert!(!response.has_single_winning_outcome());

        response.resolution = Some(scalar_resolution());
        assert!(response.is_resolved());
        assert_eq!(response.single_winning_outcome(), None);
        assert!(!response.has_single_winning_outcome());

        response.resolution = Some(single_winner_resolution());
        assert!(response.is_resolved());
        assert_eq!(response.single_winning_outcome(), Some(1));
        assert!(response.has_single_winning_outcome());
    }

    #[test]
    fn market_conversion_preserves_scalar_resolution() {
        let market = Market::try_from(valid_market_response(Some(scalar_resolution()))).unwrap();

        assert!(market.is_resolved());
        assert_eq!(market.single_winning_outcome(), None);
        assert!(!market.has_single_winning_outcome());
        let resolution = market.resolution.unwrap();
        assert_eq!(resolution.kind, wire::MarketResolutionKind::Scalar);
        assert_eq!(resolution.payout_denominator, 10);
        assert_eq!(resolution.payouts.len(), 2);
    }

    #[test]
    fn market_conversion_preserves_single_winner_resolution() {
        let market =
            Market::try_from(valid_market_response(Some(single_winner_resolution()))).unwrap();

        assert!(market.is_resolved());
        assert_eq!(market.single_winning_outcome(), Some(1));
        assert!(market.has_single_winning_outcome());
    }

    #[test]
    fn test_market_missing_slug_fails() {
        let mut resp = minimal_market_response();
        resp.slug = None;
        let err = Market::try_from(resp).unwrap_err();
        assert!(format!("{err}").contains("slug") || format!("{err}").contains("MissingSlug"));
    }

    #[test]
    fn test_market_missing_name_fails() {
        let mut resp = minimal_market_response();
        resp.market_name = None;
        let err = Market::try_from(resp).unwrap_err();
        assert!(
            format!("{err}").contains("name") || format!("{err}").contains("MarketNameMissing")
        );
    }

    #[test]
    fn test_market_invalid_status_fails() {
        let mut resp = minimal_market_response();
        resp.market_status = "UnknownStatus".to_string();
        let result = Market::try_from(resp);
        // Should fail with InvalidStatus in the error chain
        assert!(result.is_err());
    }

    #[test]
    fn test_market_missing_deposit_asset_pairs_fails() {
        // A response with no deposit assets and no orderbooks yields zero pairs.
        let resp = minimal_market_response();
        let err = Market::try_from(resp).unwrap_err();
        assert!(
            format!("{err}").contains("Missing deposit asset pairs"),
            "expected MissingDepositAssetPairs in error, got: {err}",
        );
    }

    #[test]
    fn deposit_asset_pairs_are_sorted_by_display_priority() {
        use tokens::Token;
        let pairs = super::derive_deposit_asset_pairs(
            &[
                deposit_asset("USDC"),
                deposit_asset("SOL"),
                deposit_asset("WETH"),
                deposit_asset("AAA"),
                deposit_asset("WBTC"),
                deposit_asset("ETH"),
                deposit_asset("BTC"),
                deposit_asset("ZZZ"),
            ],
            &[
                orderbook_pair("BTC", "USDC", 0),
                orderbook_pair("WBTC", "USDC", 0),
                orderbook_pair("ETH", "USDC", 0),
                orderbook_pair("WETH", "USDC", 0),
                orderbook_pair("SOL", "USDC", 0),
                orderbook_pair("AAA", "USDC", 0),
                orderbook_pair("ZZZ", "USDC", 0),
            ],
        );

        let sorted = sort_by_display_priority(&pairs);
        let base_symbols: Vec<&str> = sorted.iter().map(|pair| pair.base.symbol()).collect();
        assert_eq!(
            base_symbols,
            vec!["BTC", "WBTC", "ETH", "WETH", "SOL", "AAA", "ZZZ"]
        );
    }

    #[test]
    fn sort_by_display_priority_accepts_orderbook_pairs() {
        // Type-check sanity: `OrderBookPair: HasDisplayToken` lets the same sort
        // helper accept a list of pairs. Fixtures here all share a symbol so we
        // only assert the call type-checks and returns the expected count;
        // meaningful ordering is covered by
        // `deposit_asset_pairs_are_sorted_by_display_priority`.
        let pairs = vec![
            orderbook_pair("BTC", "DAI", 0),
            orderbook_pair("ETH", "DAI", 0),
        ];

        let sorted = sort_by_display_priority(&pairs);
        assert_eq!(sorted.len(), 2);
    }

    fn deposit_asset(mint: &str) -> tokens::DepositAsset {
        tokens::DepositAsset {
            id: 1,
            market_pda: PubkeyStr::from("market".to_string()),
            deposit_asset: PubkeyStr::from(mint.to_string()),
            num_outcomes: 2,
            name: mint.to_string(),
            symbol: mint.to_string(),
            description: None,
            decimals: 6,
            icon_url_low: String::new(),
            icon_url_medium: String::new(),
            icon_url_high: String::new(),
        }
    }

    fn orderbook_pair(
        base_mint: &str,
        quote_mint: &str,
        outcome_index: i16,
    ) -> orderbook::OrderBookPair {
        use crate::shared::OrderBookId;
        orderbook::OrderBookPair {
            id: outcome_index as i32,
            market_pubkey: PubkeyStr::from("market".to_string()),
            orderbook_id: OrderBookId::from(format!("ob-{outcome_index}")),
            base: tokens::ConditionalToken::test_new_with_deposit(
                format!("cond-base-{outcome_index}"),
                outcome_index,
                base_mint,
            ),
            quote: tokens::ConditionalToken::test_new_with_deposit(
                format!("cond-quote-{outcome_index}"),
                outcome_index,
                quote_mint,
            ),
            outcome_index,
            tick_size: 1,
            total_bids: 0,
            total_asks: 0,
            last_trade_price: None,
            last_trade_time: None,
            active: true,
        }
    }

    #[test]
    fn derive_deposit_asset_pairs_deduplicates_across_outcomes() {
        let base = deposit_asset("USDC");
        let quote = deposit_asset("USDT");
        let pairs = super::derive_deposit_asset_pairs(
            &[base.clone(), quote.clone()],
            &[
                orderbook_pair("USDC", "USDT", 0),
                orderbook_pair("USDC", "USDT", 1),
            ],
        );

        assert_eq!(pairs.len(), 1);
        assert_eq!(pairs[0].id, "USDC-USDT");
        assert_eq!(pairs[0].base, base);
        assert_eq!(pairs[0].quote, quote);
    }

    #[test]
    fn derive_deposit_asset_pairs_skips_orderbook_pairs_without_matching_deposit_asset() {
        let base = deposit_asset("USDC");
        let pairs =
            super::derive_deposit_asset_pairs(&[base], &[orderbook_pair("USDC", "MISSING", 0)]);

        assert!(pairs.is_empty());
    }

    #[test]
    fn derive_deposit_asset_pairs_returns_all_distinct_pairs() {
        let usdc = deposit_asset("USDC");
        let usdt = deposit_asset("USDT");
        let dai = deposit_asset("DAI");
        let mut pairs = super::derive_deposit_asset_pairs(
            &[usdc, usdt, dai],
            &[
                orderbook_pair("USDC", "USDT", 0),
                orderbook_pair("USDC", "DAI", 0),
            ],
        );
        pairs.sort_by(|a, b| a.id.cmp(&b.id));

        assert_eq!(pairs.len(), 2);
        assert_eq!(pairs[0].id, "USDC-DAI");
        assert_eq!(pairs[1].id, "USDC-USDT");
    }
}
